### PR TITLE
Less @internal API 

### PR DESCRIPTION
**Pull request by OliverSkroblin**
### 1. Why is this change necessary?
Some of the classes changed in the pull request are now already used by developers to extend the system. However, the classes are marked with @internal which can lead to confusion. Furthermore, classes marked with @internal are not covered by the BC Checker tool. However, since it is Public API, these classes should not be excluded.

If you know of any other classes, feel free to add them as a comment or suggestion and I will add them.

### CODE REVIEW

File: src/Core/Checkout/Cart/RuleLoader.php [hunk 1]
```diff
@@ -12,12 +12,12 @@
 use Shopware\Core\Framework\Log\Package;
 use Shopware\Core\Framework\Plugin\Exception\DecorationPatternException;
 
-/**
- * @internal
```
Comment (keulinho) [inline src/Core/Checkout/Cart/RuleLoader.php:16]: IMHO this should stay internal, the abstract class is not internal anymore (with this PR) nobody should use and depend on this class directly, so we can evolve it forward and keep backward compatibility on the abstract class level

File: src/Core/Content/Flow/Dispatching/AbstractFlowLoader.php [hunk 2]
```diff
@@ -4,9 +4,6 @@
 
 use Shopware\Core\Framework\Log\Package;
 
-/**
- * @internal not intended for decoration or replacement
```
Comment (keulinho) [inline src/Core/Content/Flow/Dispatching/AbstractFlowLoader.php]: what changed here that we now think it is a good idea to make the API public?

There are enough extension points in the flow builder itself, so what is the use case for this needing to be public?

File: src/Core/Content/Flow/Dispatching/Action/AddCustomerAffiliateAndCampaignCodeAction.php [hunk 3]
```diff
@@ -12,7 +12,7 @@
 use Shopware\Core\Framework\Uuid\Uuid;
 
 /**
- * @internal
```
Comment (keulinho) [inline src/Core/Content/Flow/Dispatching/Action/AddCustomerAffiliateAndCampaignCodeAction.php]: All the actions where actually intended to be `@internal`, you should not rely on any single concrete action, the abstract `FlowAction` is public and that should be the area where we keep backwards compatibility

File: src/Core/Framework/App/Command/ActivateAppCommand.php [hunk 4]
```diff
@@ -9,7 +9,7 @@
 use Symfony\Component\Console\Attribute\AsCommand;
 
 /**
- * @internal only for use by the app-system, will be considered internal from v6.4.0 onward
+ * @final only for use by the app-system, will be considered internal from v6.4.0 onward
```
Comment (keulinho) [inline src/Core/Framework/App/Command/ActivateAppCommand.php]: Commands should also all be internal, we don't want to keep backwards compatibility on the code level, but we want to keep the backward compatibility on the CLI level, as that's how you should execute the commands and not per php

File: src/Core/Framework/App/Exception/ActionNotFoundException.php [hunk 5]
```diff
@@ -7,7 +7,7 @@
 use Symfony\Component\HttpFoundation\Response;
 
 /**
- * @internal only for use by the app-system, will be considered internal from v6.4.0 onward
+ * @final only for use by the app-system, will be considered internal from v6.4.0 onward
```
Comment (keulinho) [inline src/Core/Framework/App/Exception/ActionNotFoundException.php]: I'm not sure about the rule about exceptions here, IMHO there is also the case for saying they are internal and you can rely on the base class :thinking:

Comment (keulinho) [reply src/Core/Framework/App/Exception/ActionNotFoundException.php]: especially in the app namespace all exceptions should be internal, as the whole module should stay internal

File: src/Core/Framework/DataAbstractionLayer/Dbal/Common/IterableQuery.php [hunk 6]
```diff
@@ -23,7 +21,7 @@ public function fetchCount(): int;
     public function getQuery(): QueryBuilder;
 
     /**
-     * @return Offset
```
Comment (keulinho) [inline src/Core/Framework/DataAbstractionLayer/Dbal/Common/IterableQuery.php]: can we split this from the other changes? Those are two different things, and i wanna understand the root cause here, as in theory the doc annotations should have no effect for serialization where we type hint array natively

Also this change is again easy to revert and thus accidentally breaking serialization again

File: src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php [hunk 7]
```diff
@@ -17,10 +16,6 @@
 
 /**
  * @final
- *
- * @internal
```
Comment (keulinho) [inline src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php:21]: the real solution here would be to split the class in the internal message handler and a public registry right?

File: src/Core/Framework/Increment/IncrementGatewayRegistry.php [hunk 8]
```diff
@@ -6,7 +6,7 @@
 use Shopware\Core\Framework\Log\Package;
 
 /**
- * @internal - Used internally for Increment pattern
```
Comment (keulinho) [inline src/Core/Framework/Increment/IncrementGatewayRegistry.php:9]: should stay internal, nothing changed here, there are other extension patterns to hook into this

File: src/Core/Framework/Script/Api/ScriptApiRoute.php [hunk 9]
```diff
@@ -16,12 +16,15 @@
 use Symfony\Component\Routing\Annotation\Route;
 
 /**
- * @internal
```
Comment (keulinho) [inline src/Core/Framework/Script/Api/ScriptApiRoute.php]: all route classes should be internal IMHO, we keep backwards compatibility on the HTTP-API level and not necessarily on the php class level

File: src/Core/System/CustomEntity/Api/CustomEntityApiController.php [hunk 10]
```diff
@@ -13,7 +13,7 @@
 use Symfony\Component\Routing\Annotation\Route;
 
 /**
- * @internal
```
Comment (keulinho) [inline src/Core/System/CustomEntity/Api/CustomEntityApiController.php]: Same here Route/Controller classes should be internal IMHO

File: src/Core/Checkout/Cart/RuleLoader.php (hunk 1 above)
Comment (OliverSkroblin) [reply src/Core/Checkout/Cart/RuleLoader.php:16]: :+1: i will add @internal to the implementation, and the abstract has no @internal anymore to signalize that it can be "consumed"

File: src/Core/Content/Flow/Dispatching/Action/AddCustomerAffiliateAndCampaignCodeAction.php (hunk 3 above)
Comment (OliverSkroblin) [reply src/Core/Content/Flow/Dispatching/Action/AddCustomerAffiliateAndCampaignCodeAction.php]: And you see no usage of directly calling the actions?

File: src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php (hunk 7 above)
Comment (OliverSkroblin) [reply src/Core/Framework/DataAbstractionLayer/Indexing/EntityIndexerRegistry.php:21]: i will mark the __invoke method as @internal

File: src/Core/Framework/Increment/IncrementGatewayRegistry.php (hunk 8 above)
Comment (OliverSkroblin) [reply src/Core/Framework/Increment/IncrementGatewayRegistry.php:9]: but you are not able to get access to a configured pool when you do not use this registry. Why having a registry when saying "use the direct implementation"

File: phpstan.neon.dist [hunk 11]
```diff
@@ -164,13 +164,6 @@ parameters:
                 - src/Core/Checkout/Order/Aggregate/OrderDelivery/OrderDeliveryDefinition.php
                 - src/Core/System/NumberRange/NumberRangeDefinition.php
 
-        - # NEXT-25276 - Needs to be fixed with the update ot PHPUnit 10
```
Comment (mitelg) [inline phpstan.neon.dist]: should not be removed. as this is not related to the changes right? also PHPStan will show errors, if you are on current trunk :wink:

**Comment by netzperfekt** [comment]
could you remove @internal from Shopware\Core\Framework\App\AppEntity ?
we have a plugin which logs the installation/activation/... of entities system wide and this includes also Apps
otherwise this plugin will not continue to work in sw 6.5... (PluginEntity is not marked as @internal...)

**Comment by netzperfekt** [comment]
and some more...

- EntityIndexingMessage
- JsonFieldSerializer (needed to generate cache keys)
- OrderLineItemEntity->getPromotion() / ->getPromotionId()
- DocumentConfigLoader

File: src/Core/DevOps/StaticAnalyze/PHPStan/Rules/Internal/InternalClassRule.php [hunk 12]
```diff
@@ -31,7 +39,20 @@ class InternalClassRule implements Rule
 
     private const INTERNAL_NAMESPACES = [
         '\\DevOps\\StaticAnalyze',
-        '\\Framework\\Demodata',
+    ];
+    private const SUBSCRIBER_EXCEPTIONS = [
+        RefreshIndexCommand::class,
```
Comment (keulinho) [inline src/Core/DevOps/StaticAnalyze/PHPStan/Rules/Internal/InternalClassRule.php:44]: this exception is not needed i guess

Comment (netzperfekt) [reply src/Core/DevOps/StaticAnalyze/PHPStan/Rules/Internal/InternalClassRule.php:44]: ok, what about the others (JsonFieldSerializer, OrderLineItemEntity->getPromotion() / ->getPromotionId()
DocumentConfigLoader)

Comment (keulinho) [reply src/Core/DevOps/StaticAnalyze/PHPStan/Rules/Internal/InternalClassRule.php:44]: For:
`OrderLineItemEntity->getPromotion() / ->getPromotionId()`
I don't see why they are internal we should make them public IMHO

`DocumentConfigLoader`
We should split the actual loading and the "caching" and resetting in two services, the actual loading should be public (either by adding a public abstract class or just remove internal from the actual loader class)

`JsonFieldSerializer`
All serializer are internal, this is not public api for us, and i would like to keep it that way, can you explain the use case why you need to access that class? Which cache keys to you need to generate and how?

Comment (netzperfekt) [reply src/Core/DevOps/StaticAnalyze/PHPStan/Rules/Internal/InternalClassRule.php:44]: @keulinho `JsonFieldSerializer` - look at your own docs: https://developer.shopware.com/docs/guides/plugins/plugins/framework/store-api/add-caching-for-store-api-route
(generateKey() at the very end)

of course, the current implementation of `encodeJson` is quite trivial and i could just copy that into my code. but that doesn't feel right.

File: phpstan.neon.dist (hunk 11 above)
Comment (OliverSkroblin) [reply phpstan.neon.dist]: reverted

File: src/Core/DevOps/StaticAnalyze/PHPStan/Rules/Internal/InternalClassRule.php (hunk 12 above)
Comment (OliverSkroblin) [reply src/Core/DevOps/StaticAnalyze/PHPStan/Rules/Internal/InternalClassRule.php:44]: @netzperfekt I adjusted some code fragments you mentioned:

- EntityIndexingMessage: __construct is no more `@internal`
- JsonFieldSerializer: I implemented a new Json class which can be used - docs update follows
- OrderLineItemEntity->getPromotion(): Is no more `@internal`
- DocumentConfigLoader - Is now marked as public API, but only the `load` function to allow calling the service.
- AppEntity: is no more `@internal`

**Comment by shyim** [comment]
merged with https://github.com/shopware/platform/commit/b165aee50c0205595e874828b73495e81b105f9f
